frontend/src/metabase/reference/selectors.js

```diff
@@ -0,0 +1,475 @@
+const getDatabase = createSelector(
+    [getDatabaseId, getDatabases],
+    (databaseId, databases) => databases[databaseId] || { id: databaseId }
+);
+
+export const getTableId = (state) => state.router.params.tableId;
+export const getTables = (state) => state.metadata.tables;
+const getTablesByDatabase = createSelector(
+    [getTables, getDatabase],
+    (tables, database) => tables && database ? Object.values(tables)
+        .filter(table => table.db_id === database.id)
+        .reduce((tableMap, table) => i.assoc(tableMap, table.id, table), {}) : {}
+);
+
+const getTable = createSelector(
+    [getTableId, getTables],
+    (tableId, tables) => tables[tableId] || { id: tableId }
+);
+const getTableByList = createSelector(
+    [getList, getTables],
+    (list, tables) => list ? tables[list.table_id] : {}
+);
+
+export const getFieldId = (state) => state.router.params.fieldId;
+const getFields = createSelector(
+    [getTable],
+    (table) => table && table.fields ?
+        table.fields.reduce((fieldMap, field) => i.assoc(fieldMap, field.id, field), {}) : {}
+);
+const getFieldsByList = createSelector(
+    [getTableByList],
+    (table) => table && table.fields ?
+        table.fields.reduce((fieldMap, field) => i.assoc(fieldMap, field.id, field), {}) : {}
+);
+const getField = createSelector(
+    [getFieldId, getFields],
+    (fieldId, fields) => fields[fieldId] || { id: fieldId }
+);
+const getFieldByList = createSelector(
+    [getFieldId, getFieldsByList],
+    (fieldId, fields) => fields[fieldId] || { id: fieldId }
+);
+
+const getQuestions = (state) => i.getIn(state, ['questions', 'entities', 'cards']) || {};
+// TODO: get someone to look over this
+// probably not be the best way to determine which questions are associated
+// with a particular metric/list, but seems to work
+const filterMetricQuestions = (metricId, question) => {
+    const aggregation = i.getIn(question, ['dataset_query', 'query', 'aggregation']);
+    if (!aggregation) {
+        return false;
+    }
+
+    return aggregation[0] === "METRIC" && aggregation[1].toString() === metricId;
+};
+
+const getMetricQuestions = createSelector(
+    [getMetricId, getQuestions],
+    (metricId, questions) => Object.values(questions)
+        .filter(question => filterMetricQuestions(metricId, question))
+        .reduce((map, question) => i.assoc(map, question.id, question), {})
+);
+
+const getRevisions = (state) => state.metadata.revisions;
+
+const getMetricRevisions = createSelector(
+    [getMetricId, getRevisions],
+    (metricId, revisions) => i.getIn(revisions, ['metric', metricId]) || {}
+);
+
+const getListRevisions = createSelector(
+    [getListId, getRevisions],
+    (listId, revisions) => i.getIn(revisions, ['list', listId]) || {}
+);
+
+const filterListQuestions = (listId, question) => {
+    const filter = i.getIn(question, ['dataset_query', 'query', 'filter']);
+    if (!filter) {
+        return false;
+    }
+    if (filter[0] === "AND") {
```

Comment:
When manipulating `dataset_query` you should try to use the functions in `metabase/lib/query.js`, e.x. `getFilters` will return a list of filters regardless of whether it's wrapped in `[AND ...]`, `isSegmentFilter` can be used to check if a filter is a segment, etc.